jerry-core/parser/js/js-parser-module.c

```diff
@@ -0,0 +1,803 @@
+{
+  parser_module_context_t *module_context_p = context_p->module_context_p;
+
+  if (module_context_p == NULL)
+  {
+    return;
+  }
+
+  parser_module_node_t *current_node_p = module_context_p->imports_p;
+  parser_module_node_t *next_node_p;
+
+  while (current_node_p != NULL)
+  {
+    parser_free (current_node_p->script_path_p, current_node_p->script_path_length * sizeof (uint8_t));
+    parser_module_delete_saved_names (current_node_p);
+
+    next_node_p = current_node_p->next_p;
+
+    parser_free (current_node_p, sizeof (parser_module_node_t));
+    current_node_p = next_node_p;
+  }
+
+  parser_free (module_context_p, sizeof (parser_module_context_t));
+  context_p->module_context_p = NULL;
+} /* parser_module_cleanup_module_context */
+
+/**
+ * Create module context and bind to the parser context.
+ */
+void
+parser_module_context_init (parser_context_t *context_p) /**< parser context */
+{
+  if (context_p->module_context_p == NULL)
+  {
+    context_p->module_context_p =
+    (parser_module_context_t *) parser_malloc (context_p, sizeof (parser_module_context_t));
+
+    context_p->module_context_p->exports_p = NULL;
+    context_p->module_context_p->imports_p = NULL;
+  }
+} /* parser_module_context_init */
+
+/**
+ * Create import node.
+ * @return the created import node
+ */
+parser_module_node_t *
+parser_module_create_import_node (parser_context_t *context_p, /**< parser context */
+                                  parser_module_node_t *template_node_p) /**< template node for the new node */
+{
+  parser_module_node_t *node = (parser_module_node_t *) parser_malloc (context_p, sizeof (parser_module_node_t));
+
+  bool has_template = template_node_p != NULL;
+  node->module_names_p = has_template ? template_node_p->module_names_p : NULL;
+  node->module_request_count = has_template ? template_node_p->module_request_count : 0;
+
+  node->script_path_p = has_template ? template_node_p->script_path_p : NULL;
+  node->script_path_length = has_template ? template_node_p->script_path_length : 0;
+
+  node->next_p = NULL;
+
+  return node;
+} /* parser_module_create_import_node */
+
+/**
+ * Create export node or get the previously created one.
+ * @return the export node
+ */
+parser_module_node_t *
+parser_module_get_export_node (parser_context_t *context_p) /**< parser context */
+{
+  if (context_p->module_context_p->exports_p != NULL)
+  {
+    return context_p->module_context_p->exports_p;
+  }
+
+  return parser_module_create_import_node (context_p, NULL);
+} /* parser_module_get_export_node */
+
+/**
+ * Parse export item list.
+ */
+void
+parser_module_parse_export_item_list (parser_context_t *context_p, /**< parser context */
+                                      parser_module_node_t *module_node_p) /**< module node */
+{
+  if (context_p->token.type == LEXER_LITERAL
+      && lexer_compare_raw_identifier_to_current (context_p, "from", 4))
+  {
+    parser_raise_error (context_p, PARSER_ERR_INVALID_CHARACTER);
```

Comment:
If an error is raised, whether the initialized context is freed? (This should be done in the `PARSER_CATCH` part of js-parser.c